drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,960 @@
+		    *reg_data);
+	axi_io_read(desc->rx_dma_baseaddr, reg_addr, reg_data);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set width of the transfered word over SPI
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param data_wdith The desired data width
+ * 	The supported values are:
+ * 		- 8
+ * 		- 16
+ * 		- 24
+ * 		- 32
+ */
+static int32_t spi_engine_set_transfer_wdith(struct spi_desc *desc,
+		uint8_t data_wdith)
+{
+	struct spi_engine_desc	*desc_extra;
+
+	desc_extra = desc->extra;
+
+	if (data_wdith > desc_extra->max_data_width) {
+		desc_extra->data_width = desc_extra->max_data_width;
+		dev_dbg(desc, "Data width (%d) exceeds the maximum of %d bits",
+			data_wdith, desc_extra->max_data_width);
+	} else {
+		dev_dbg(desc, "Data width set to %d bits",data_wdith);
+		desc_extra->data_width = data_wdith;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set the number of words transfered in a single transaction
+ *
+ * @param desc Decriptor containing SPI Engine's parameters
+ * @param bytes_number The number of bytes to be cnoverted
+ * @return uint8_t A number of words in which bytes_number can be grouped
+ */
+static uint8_t spi_get_words_number(struct spi_engine_desc *desc,
+				    uint8_t bytes_number)
+{
+	uint8_t xfer_word_len;
+	uint8_t words_number;
+
+	xfer_word_len = desc->data_width / 8;
+	words_number = bytes_number / xfer_word_len;
+
+	if ((bytes_number % xfer_word_len) != 0)
+		words_number++;
+
+	return words_number;
+}
+
+/**
+ * @brief Get the word lenght in bytes
+ *
+ * @param desc Decriptor containing SPI Engine's parameters
+ * @return uint8_t Number of bytes that fit in one word
+ */
+static uint8_t spi_get_word_lenght(struct spi_engine_desc *desc)
+{
+	uint8_t word_lenght;
+
+	word_lenght = desc->data_width / 8;
+
+	return word_lenght;
+}
+
+/**
+ * @brief Compute the prescaler used to set the sleep period.
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param sleep_time_ns The amount of time where the transfer hangs
+ * @param sleep_div Clock prescaler
+ * @return int32_t This function allways returns SUCCESS
+ */
+static int32_t spi_get_sleep_div(struct spi_desc *desc,
+				 uint32_t sleep_time_ns,
+				 uint32_t *sleep_div)
+{
+	struct spi_engine_desc	*eng_desc;
+
+	eng_desc = desc->extra;
+
+	/*
+	 * Engine Wiki:
+	 *
+	 * The frequency of the SCLK signal is derived from the
+	 * module clock frequency using the following formula:
+	 * f_sclk = f_clk / ((div + 1) * 2)
+	 */
+
+	*sleep_div = (desc->max_speed_hz / 1000000 * sleep_time_ns / 1000) /
+		     ((eng_desc->clk_div + 1) * 2) - 1;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Create a new commands queue used in a spi transfer
+ *
+ * @param fifo Command buffer, ussualy used in fifo mode
+ * @param cmd Command with wich the buffer will be initiated
+ * @return int32_t Failure if the memory allocation failed
+ */
+static int32_t spi_engine_queue_new_cmd(struct spi_engine_cmd_queue **fifo,
```

Comment:
nitpick: the command queue here behaves more like a directed graph than a FIFO, so probably another parameter name would be appropriate.

Reply:
This queue has the possibility to add and remove items in both fifo and lifo modes. The name was chosen because in most cases is used as a fifo.